I624 validator check for duplicate ids

kerimoyle opened the pull request:
Addresses #624 and #635 

At present this checks the model scope for duplicate ids, because this is - in effect - no different from checking a parsed document.

Changes in this PR:
- ~added internal utility function to retrieve `id` attribute from `math` strings,~
- ~added functions to API to get id for `Component::mathId()`, `Reset::testValueMathId()`, `Reset::resetValueMathId()`,~
- added tests for new functions,
- added tests for IDs on all items, not just `Entity` items,
- added function to `Validator` to check for duplicate ids in the model and report them.
- added functionality to parse MathML strings in components and resets and validate uniqueness.
- changed minimum compiler to C++17 (sorry, I know it's in PR #636 but it's here too ...)

kerimoyle committed:
Added parser test for math with id

kerimoyle committed:
Validating unique IDs across model scope.  WIP depending on issue #629

kerimoyle committed:
Add test to demonstrate usage of proposed mathId() function. Will not compile.

kerimoyle commented on src/validator.cpp:
```diff
@@ -1355,4 +1389,178 @@ void Validator::ValidatorImpl::checkUnitForCycles(const ModelPtr &model, const U
+            auto equiv = item->equivalentVariable(e);
+            if (equiv != nullptr) {
+                auto equivParent = owningComponent(equiv);
+                if (equivParent != nullptr) {
```
@agarny This is where I used them ... now there are checks, but I think it would be cleaner to be able to pass in a nullptr to owningComponent and owningModel and have it return a null rather than a segfault?

agarny replied:
Sure, but that's besides the point. `owningComponent()` can't be tested with `nullptr` as a parameter. Our tests only have access to the API, not our utilities.

kerimoyle replied:
I think we don't understand each other.  I would rather not have these two nested if test in my PR version - I'd rather be able to call owningComponent() safely.  This means that if I change the owningComponent function in my PR, and also change my code back to how I wanted it, it will be called during testing, cos that's how I found the segfault in the first place.  It doesn't appear in your PR because your PR doesn't have these functions to check ID.  So coverage will be fine ... ?

agarny replied:
Oh, ok, got it. Sure, if your code can somehow result in `owningModel()` and `owningComponent()` to be called with `nullptr` as a parameter, then by all means update it in your PR.

(Phew, we are getting there!  Too much noise around me at the office. Am not used to it anymore... and don't want to get used to it again!)

kerimoyle committed:
Added utility function to get id attribute from math blocks

kerimoyle committed:
Updated validator to check MathML ids.

kerimoyle committed:
Added math to model string to validate for duplicate ids

kerimoyle committed:
Added parser tests for non-entity items (connections, mappings, maths, encaps)

kerimoyle committed:
Removed test that's now in #632

kerimoyle committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

kerimoyle committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

kerimoyle committed:
Fixed test for ids of all items

kerimoyle committed:
Merge branch 'i624-Validator-check-for-duplicate-ids' of https://github.com/kerimoyle/libcellml into i624-Validator-check-for-duplicate-ids

kerimoyle committed:
Docstrings

kerimoyle committed:
make mathId const

kerimoyle committed:
Coverage: removed unreachable code.

kerimoyle commented:
Ready for review please @agarny @hsorby @nickerso .

kerimoyle committed:
Merge branch 'i634_c++17' into i624-Validator-check-for-duplicate-ids

kerimoyle commented:
I've added functionality to parse the MathML string for id attributes anywhere, and check them against the rest of the model.  The wording for errors uses "maths block at index 0" etc as I'd assumed we were splitting up the string for separate retrieval (as in #633).  If we're sure that's not going to make it into the API then I'll change the wording to "first", "second" etc ? @hsorby @agarny @nickerso

kerimoyle committed:
Validator tests for duplicate element ids in mathml

nickerso commented on src/api/libcellml/component.h:
```diff
@@ -122,6 +122,13 @@ class LIBCELLML_EXPORT Component: public ComponentEntity, public ImportedEntity
+     *
+     * @return id The @c std::string if the id attribute is present, or an empty string if not.
+     */
+    std::string mathId() const;
```
This should be removed as it is not clear which ID would be returned.

kerimoyle replied:
d4b98796

nickerso commented on src/validator.cpp:
```diff
@@ -291,6 +293,50 @@ struct Validator::ValidatorImpl
+     *
+     * @param model The model to be checked.
+     */
+    void checkForDuplicateIds(const ModelPtr &model);
```
for me, this should be more more like `checkUniqueIds`

kerimoyle replied:
aa2e8d22

nickerso commented on src/validator.cpp:
```diff
@@ -1355,4 +1404,231 @@ void Validator::ValidatorImpl::checkUnitForCycles(const ModelPtr &model, const U
+            addIdMapItem(item->testValueId(), info, idMap);
+        }
+        if (!item->testValueMathId().empty()) {
+            info = "  - MathML child of test_value in reset at index " + std::to_string(i) + " in component '" + component->name() + "'\n";
```
so this only checks the `math` element child for an id, not the full MathML string?

kerimoyle replied:
Ah, good point!  I'd forgotten about the resets ... will do that now.

kerimoyle replied:
aa2e8d22

kerimoyle committed:
Reset maths now tested in validator

kerimoyle committed:
removed all mathId functionality from components and resets

kerimoyle commented:
Ready for re-review please @agarny @hsorby @nickerso

nickerso committed:
Tidy up id checking comments a little.

agarny commented:
I just had a quick look at the code and, like @nickerso (from a comment on Slack, I believe), it looks overcontrived to me when it could be short and sweet using [`XPath`](https://en.wikipedia.org/wiki/XPath). I have never used `XPath`, but [`libxml2`](http://xmlsoft.org/) has an [`XPath` module](http://xmlsoft.org/html/libxml-xpath.html) and it comes with a quick [tutorial](http://xmlsoft.org/tutorial/ar01s05.html). So, it might be worth looking into it.

kerimoyle commented:
There are two parts to the new code: stuff to check the ids on items which are accessible through libCellML, and stuff to check the XML in the MathML strings.  
1) The iteration through the former means that we can return useful messages to the user.  This is to avoid the "duplicate found but I'm not telling what or where" unhelpful messages which come from libxml2 and its checks.
2) The iteration in the latter uses exactly the same logic as the parser and validator already use to read and manipulate MathML strings - I just followed what we already do.

If you can narrow down which part you're objecting to that might be helpful for me to figure out how to do it better? As it is, the choice I'm seeing is either getting awful error messages back, or redoing the current XML stuff throughout the rest of libCellML too?

agarny commented:
Based on your current approach, your code looks fine.

However, looking at your `duplicateIdAll` test, I can for instance see that `id1` is duplicated. Fine. Then, I can see several "sub" messages about where `id1` is duplicated. To be honest, to read all those messages and then find where in the XML code `id1` is duplicated (based on the message information) would take me much longer than search for `id1` in the whole XML file. Then, I would have to do the same for `id2`, `id3`, etc.

So, personally, I would just let the user know that `id1`, `id2`, `id3`, etc. are duplicated and that's all.

kerimoyle commented:
That's fine for users which have an interface (like OpenCOR) which mean that the items can be edited and searched without directly reading the XML.  But I thought that the whole point of libcellml was that users didn't need to look at the raw code - hence adding the message about which objects needed to be fixed.  
I realise that MathML is a little different in that we still treat it as a string, so users will have to create/edit that string somehow, but even that sounds like it will change in the near future anyway.  The shorter error messages (the ones you like) come from duplications inside the same MathML block (and from the validateMath function earlier in the validator), rather than across the entire document (which is what the new function adds).

kerimoyle committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

agarny committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

kerimoyle committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

hsorby committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

kerimoyle commented:
ping @hsorby @agarny @nickerso

kerimoyle committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

kerimoyle committed:
Trigger new build.

kerimoyle committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

kerimoyle committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

nickerso committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

kerimoyle committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

kerimoyle committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

kerimoyle committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

kerimoyle committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

kerimoyle committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

kerimoyle committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

kerimoyle committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

kerimoyle committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

kerimoyle committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

kerimoyle committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

kerimoyle committed:
Merge branch 'develop' into i624-Validator-check-for-duplicate-ids

agarny committed:
Some minor cleaning up.

agarny committed:
Validator: removed an unneeded test.

We are iterating through (existing) equivalent variables, so we should never have a null equivalent variable...?

agarny committed:
Some minor cleaning up.

agarny committed:
Validator: slight improvement to the reporting of duplicated ids.

agarny reviewed (dismissed):
Did a bit of cleaning up and also slightly reformatted the reporting of duplicated ids, but apart from that it looks good to me.

agarny committed:
Reset: removed an unneeded include file.

nickerso commented on src/validator.cpp:
```diff
@@ -1376,4 +1425,237 @@ void Validator::ValidatorImpl::checkUnitForCycles(const ModelPtr &model, const U
+            return;
+        }
+        if (!node->isMathmlElement("math")) {
+            return;
```
why is this return instead of ignoring non-math root nodes?

kerimoyle replied:
7f063dfa

agarny committed:
Validator test: fixed a small indentation issue.

nickerso reviewed (changes requested):
just the one little comment to address I think...

kerimoyle committed:
Checks for mathml ids continue after non-math tag

kerimoyle committed:
Merge branch 'i624-Validator-check-for-duplicate-ids' of https://github.com/kerimoyle/libcellml into i624-Validator-check-for-duplicate-ids

kerimoyle committed:
Added test

agarny committed:
Some minor cleaning up.
